site/content/features/recordpatterns.md

```diff
@@ -0,0 +1,619 @@
+      return end;
+   }
+   public String toString() { return seq.subSequence(0, end).toString(); }
+}
+
+record Final(CharSequence seq, int start) implements SubSequence {
+   public int start() {
+      Objects.checkIndex(start, seq.length());
+      return start;
+   }
+   public String toString() { return seq.subSequence(start, seq.length()).toString(); }
+}
+
+record Middle(CharSequence seq, int start, int end) implements SubSequence {
+   public int start() {
+      Objects.checkFromToIndex(start, end, seq.length());
+      return start;
+   }
+   public int end() {
+      Objects.checkFromToIndex(start, end, seq.length());
+      return end;
+   }
+   public String toString() { return seq.subSequence(start, end).toString(); }
+}
+
+public class Main {
+   public static void main(String[] args) {
+      CharSequence seq = new Middle("Mississippi", 5, 30);    
+      System.out.println(seq.length());
+   }
+}
+
+{{< /sandboxsource >}}
+{{< /sandbox >}}
+
+Rémi Forax has a [more amusing example](https://mail.openjdk.org/pipermail/amber-spec-observers/2022-June/003626.html) which generates a linked list of match errors.
+
+I think it is best not to override component accessors so that they throw exceptions. It is ok to throw an exception in the constructor when construction parameters are `null` or out of range. But once the record instance is constructed, one should be able to extract its state, no matter what it is.
+
+## How Momentous Are Record Patterns?
+
+Records are nice when they match your needs—a class that describes immutable objects that are “just data”. As of today, the JDK has one of them. Of course, they are a recent feature, so surely there will be more to come. You probably have a few classes in your code base that would work well as records, and maybe you have started declaring your own.
+
+For record patterns to be useful, you need multiple records that implement a common interface. You saw a nontrivial example with the `SubSequence` records. Another example is an expression hierarchy with records `Sum`, `Difference`, `Product`, `Quotient`. Or a functional list or tree with a record for the nodes. These examples are, depending on your point of view, foundational or academic. Either way, they are unlikely to feature prominently in your business logic.
+
+Record patterns are a piece of the pattern matching toolset that Java is building up. Useful libraries may emerge in the future that make good use of them. Even more so when Java supports value types. A JSON library would be a plausible example. Keep record patterns on your radar, but pay no attention to those who tell you that they are a crucial reason to update your JDK.
+
+## References
+
+* [JEP 394: Record Patterns (Preview), OpenJDK](https://openjdk.java.net/jeps/405)
+* [JEP 427: Pattern Matching for switch (Third Preview), OpenJDK](https://openjdk.java.net/jeps/427)
+
+Comments powered by [Talkyard](https://www.talkyard.io).
```

Comment:
Is this intentional? Looks like the relative links below do not work on javaalmanac.io

Reply:
Those weren't meant to be there. They are for the version on my blog. I removed them.